public/app/features/dashboard/dashgrid/PanelPerformanceMonitor.tsx

```diff
@@ -0,0 +1,159 @@
+import { useEffect, useRef } from 'react';
+
+import { FieldConfigSource } from '@grafana/data';
+import { faro } from '@grafana/faro-web-sdk';
+import appEvents from 'app/core/app_events';
+import { DashboardSavedEvent } from 'app/types/events';
+
+interface Props {
+  children: JSX.Element;
+  isInPanelEdit: boolean;
+  panelType: string;
+  panelId: number;
+  panelTitle: string;
+  panelOptions: any;
+  panelFieldConfig: FieldConfigSource;
+}
+
+export const PanelPerformanceMonitor = (props: Props) => {
+  const startLoadTime = performance.now();
+  const panelIdStr = String(props.panelId);
+
+  const oldPanelOptions = useRef(props.panelOptions);
+  const oldPanelFieldConfig = useRef<FieldConfigSource>(props.panelFieldConfig);
+
+  const savedFieldConfig = useRef<FieldConfigSource>(props.panelFieldConfig);
+  const savedPanelOptions = useRef(props.panelOptions);
+
+  const logChanges = () => {
+    logPanelOptionChanges(savedPanelOptions.current, oldPanelOptions.current);
+    logFieldConfigChanges(savedFieldConfig.current, oldPanelFieldConfig.current);
+
+    //set the old values to the current values for next save
+    oldPanelOptions.current = savedPanelOptions.current;
+    oldPanelFieldConfig.current = savedFieldConfig.current;
+  };
+
+  useEffect(() => {
+    savedPanelOptions.current = props.panelOptions;
+  }, [props.panelOptions]);
+
+  useEffect(() => {
+    savedFieldConfig.current = props.panelFieldConfig;
+  }, [props.panelFieldConfig]);
+
+  const logEvent = (eventName: string, newKey: string, newVal: string, oldVal?: string) => {
+    const logObj = {
+      key: newKey,
+      newValue: newVal,
+      oldValue: oldVal ?? '',
+      panelTitle: props.panelTitle,
+      panelId: panelIdStr,
+      panelType: props.panelType,
+    };
+
+    //TODO: do batch push when that becomes available
+    faro.api.pushEvent(eventName, logObj);
+  };
+
+  const logPanelOptionChanges = (panelOptions: any, oldPanelOptions: any) => {
+    for (const [key, value] of Object.entries(panelOptions)) {
+      const newValue: string = typeof value !== 'string' ? JSON.stringify(value) : value;
+      const oldValue: string = typeof value !== 'string' ? JSON.stringify(oldPanelOptions[key]) : oldPanelOptions[key];
+
+      if (oldPanelOptions[key] === undefined) {
+        logEvent('new panel option', key, newValue);
+      } else if (oldValue !== newValue) {
+        logEvent('panel option changed', key, newValue, oldValue);
+      }
+    }
+  };
+
+  const logFieldConfigChanges = (fieldConfig: FieldConfigSource, oldFieldConfig: FieldConfigSource) => {
+    // overrides are an array of objects, so stringify it all and log changes
+    // in lack of an index, we can't tell which override changed
+    if (oldFieldConfig.overrides !== fieldConfig.overrides) {
+      logEvent(
+        'field config overrides changed',
+        'overrides',
+        JSON.stringify(fieldConfig.overrides),
+        JSON.stringify(oldFieldConfig.overrides)
+      );
```

Comment:
fieldConfiug object can be REALLY large, including not only the defaults, many overrides, and hundreds of mappings